Firestore/core/src/firebase/firestore/model/timestamp.cc

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2018 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "Firestore/core/src/firebase/firestore/model/timestamp.h"
+
+#include <time.h>
+
+#include "Firestore/core/src/firebase/firestore/util/firebase_assert.h"
+
+namespace firebase {
+namespace firestore {
+namespace model {
+
+namespace {
+long kOriginSeconds = -62135596800L;
+}
+
+Timestamp::Timestamp(long seconds, int nanos)
+    : seconds_(seconds), nanos_(nanos) {
+  FIREBASE_ASSERT_MESSAGE_WITH_EXPRESSION(
+      nanos >= 0, nanos >= 0, "timestamp nanoseconds out of range: %d", nanos);
+  FIREBASE_ASSERT_MESSAGE_WITH_EXPRESSION(
+      nanos < 1e9, nanos < 1e9, "timestamp nanoseconds out of range: %d",
+      nanos);
+  // Midnight at the beginning of 1/1/1 is the earliest timestamp Firestore
+  // supports.
+  FIREBASE_ASSERT_MESSAGE_WITH_EXPRESSION(
+      seconds >= kOriginSeconds, seconds >= -62135596800L,
+      "timestamp seconds out of range: %lld", seconds);
+  // This will break in the year 10,000.
+  FIREBASE_ASSERT_MESSAGE_WITH_EXPRESSION(
+      seconds < 253402300800L, seconds < 253402300800L,
+      "timestamp seconds out of range: %lld", seconds);
+}
+
+Timestamp::Timestamp() : seconds_(time(nullptr)), nanos_(0) {
+}
+
+bool Timestamp::IsOrigin() const {
```

Comment:
The problem here in C++ is that iOS `previous_value` is `Nullable` and thus we need a dummy value for `previous_value` to present the null value. How about make a null value of timestamp that is not overlap with any true timestamp and just call the API function `bool Timestamp::IsNull()` and `static Timestamp Timestamp::NullValue()`?

Reply:
Let's reconcile this with @varconst-fb since this will be a public API.

Reply:
@zxu123, do you refer to previousValue in FSTServerTimestampValue? If yes, perhaps you could store previousValue as absl::optional, and for constructor/factory function, provide two overloads, with and without previousValue?

Reply:
I'll port abseil's optional in a separate PR. @wilhuff any comment on porting abseil's optional? The port only introduce additional:

- absl/memory/memory.{h,cc},
- absl/base/macros.{},
- absl/base/port.h,
- absl/base/attributes.,
- abs/base/config.,
- abs/base/optimization.,
- abs/utility/utility.,
- absl/base/internal/invoke.h,
- absl/types/bad_optional_access.h,

absl::optional might be also useful for other part of the C++ code due to the lack of mechanism of C++ to allow value types being null.

Reply:
I presumed we depended on Abseil, so I didn't expect this to be an issue. I guess it's possible to just have a flag inside ServerTimestamp like bool has_previous_value_. Unless there are other places which need some sort of a NaN timestamp, I'd keep this quirk inside implementation of ServerTimestamp instead of introducing a sentinel Timestamp value.

OTOH, optional will probably find other uses. I'll defer to @wilhuff on this.

Reply:
Based on these comments yesterday I started work on just doing a full import of abseil rather than doing it piecewise as we've done so far. I've come across a blocking issue I hope to work out with the abseil team today (though who knows how long it will take for my proposal to actually land in their github repo). 

For now I would add the `has_previous_value_` and a TODO so we can make progress without that dependency.